barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_gas_trace.hpp

```diff
@@ -0,0 +1,146 @@
+    { OpCode::DAGASLEFT, temp_default_gas_entry },
+    // Machine State - Internal Control Flow
+    { OpCode::JUMP, temp_default_gas_entry },
+    { OpCode::JUMPI, temp_default_gas_entry },
+    { OpCode::INTERNALCALL, temp_default_gas_entry },
+    { OpCode::INTERNALRETURN, temp_default_gas_entry },
+    // Machine State - Memory
+    { OpCode::SET, temp_default_gas_entry },
+    { OpCode::MOV, temp_default_gas_entry },
+    { OpCode::CMOV, temp_default_gas_entry },
+
+    // World State
+    { OpCode::SLOAD, temp_default_gas_entry },
+    { OpCode::SSTORE, temp_default_gas_entry },
+    { OpCode::NOTEHASHEXISTS, temp_default_gas_entry },
+    { OpCode::EMITNOTEHASH, temp_default_gas_entry },
+    { OpCode::NULLIFIEREXISTS, temp_default_gas_entry },
+    { OpCode::EMITNULLIFIER, temp_default_gas_entry },
+    { OpCode::L1TOL2MSGEXISTS, temp_default_gas_entry },
+    { OpCode::HEADERMEMBER, temp_default_gas_entry },
+    { OpCode::GETCONTRACTINSTANCE, temp_default_gas_entry },
+
+    // Accrued Substate
+    { OpCode::EMITUNENCRYPTEDLOG, temp_default_gas_entry },
+    { OpCode::SENDL2TOL1MSG, temp_default_gas_entry },
+
+    // Control Flow - Contract Calls
+    { OpCode::CALL, temp_default_gas_entry },
+    { OpCode::STATICCALL, temp_default_gas_entry },
+    { OpCode::DELEGATECALL, temp_default_gas_entry },
+    { OpCode::RETURN, temp_default_gas_entry },
+    { OpCode::REVERT, temp_default_gas_entry },
+
+    // Misc
+    { OpCode::DEBUGLOG, temp_default_gas_entry },
+
+    // Gadgets
+    { OpCode::KECCAK, temp_default_gas_entry },
+    { OpCode::POSEIDON2, temp_default_gas_entry },
+    { OpCode::SHA256, temp_default_gas_entry },
+    { OpCode::PEDERSEN, temp_default_gas_entry },
+
+    // Conversions
+    { OpCode::TORADIXLE, temp_default_gas_entry },
+
+    // Future Gadgets -- pending changes in noir
+    { OpCode::SHA256COMPRESSION, temp_default_gas_entry },
+    { OpCode::KECCAKF1600, temp_default_gas_entry }, // Here for when we eventually support this
+    // Sentinel
+    // LAST_OPCODE_SENTINEL,
+};
+class AvmGasTraceBuilder {
+  public:
+    struct GasTraceEntry {
+        uint32_t clk = 0;
+        OpCode opcode = OpCode::ADD; // 0
+        uint32_t l2_gas_cost = 0;
+        uint32_t da_gas_cost = 0;
+        uint32_t remaining_l2_gas = 0;
+        uint32_t remaining_da_gas = 0;
+    };
+
+    // Counts each time an opcode is read
+    // opcode -> count
+    std::unordered_map<OpCode, uint32_t> gas_opcode_lookup_counter;
+
+    // Constructor receives copy of kernel_inputs from the main trace builder
+    AvmGasTraceBuilder();
+
+    void reset();
+    std::vector<GasTraceEntry> finalize();
+
+    void constrain_gas_lookup(uint32_t clk, OpCode opcode);
+    void set_initial_gas(uint32_t l2_gas, uint32_t da_gas);
+
+    std::vector<GasTraceEntry> gas_trace;
+
+    uint32_t initial_l2_gas = 0;
+    uint32_t initial_da_gas = 0;
+
+  private:
+    std::vector<GasTableEntry> construct_gas_lookup_table();
```

Comment:
Is this ever used since we have the hardcoded table above?

Reply:
Good catch!